broker/fragment/store_azure.go

```diff
@@ -29,30 +35,63 @@ type azureBackend struct {
 	endpoint    string
 	accountName string
 	client      pipeline.Pipeline
-	credentials *azblob.SharedKeyCredential
+	svcClient   service.Client
 	clientMu    sync.Mutex
+	udc         *service.UserDelegationCredential
+	udc_exp     *time.Time
 }
 
 func (a *azureBackend) Provider() string {
 	return "azure"
 }
 
+// Cache UserDelegationCredentials and refresh them when needed
+func (a *azureBackend) GetUserDelegationCredential() (*service.UserDelegationCredential, error) {
+	if a.udc == nil || (a.udc_exp != nil && a.udc_exp.After(time.Now().Add(time.Minute*10))) {
+		expTime := time.Now().Add(time.Hour * 24)
+		info := service.KeyInfo{
+			Start:  to.Ptr(time.Now().Add(time.Second * -10).UTC().Format(sas.TimeFormat)),
+			Expiry: to.Ptr(expTime.UTC().Format(sas.TimeFormat)),
+		}
+
+		udc, err := a.svcClient.GetUserDelegationCredential(context.Background(), info, nil)
+
+		if err != nil {
+			return nil, err
+		}
+
+		a.udc = udc
+		a.udc_exp = &expTime
+	}
+
+	return a.udc, nil
+}
+
+// https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/storage/azblob/service/examples_test.go#L285
```

Comment:
nit: when dropping URLs like this, include _why_ the URL is there. Give me, as a reader, some sense of what I might find there.

Reply:
Added a comment